server/data/data-api/src/main/java/org/apache/james/dlp/eventsourcing/EventSourcingDLPRuleStore.java

```diff
@@ -0,0 +1,104 @@
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.dlp.eventsourcing;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import javax.inject.Inject;
+
+import org.apache.james.core.Domain;
+import org.apache.james.dlp.api.DLPRule;
+import org.apache.james.dlp.api.DLPRuleId;
+import org.apache.james.dlp.api.DLPRulesStore;
+import org.apache.james.dlp.eventsourcing.aggregates.DLPRuleAggregateId;
+import org.apache.james.dlp.eventsourcing.commands.ClearCommand;
+import org.apache.james.dlp.eventsourcing.commands.ClearCommandHandler;
+import org.apache.james.dlp.eventsourcing.commands.StoreCommand;
+import org.apache.james.dlp.eventsourcing.commands.StoreCommandHandler;
+import org.apache.james.dlp.eventsourcing.events.StoreEvent;
+import org.apache.james.eventsourcing.Event;
+import org.apache.james.eventsourcing.EventSourcingSystem;
+import org.apache.james.eventsourcing.Subscriber;
+import org.apache.james.eventsourcing.eventstore.EventStore;
+import org.apache.james.eventsourcing.eventstore.History;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+public class EventSourcingDLPRuleStore implements DLPRulesStore {
+
+    private static final ImmutableSet<Subscriber> NO_SUBSCRIBER = ImmutableSet.of();
+
+    private final EventSourcingSystem eventSourcingSystem;
+    private final EventStore eventStore;
+
+    @Inject
+    public EventSourcingDLPRuleStore(EventStore eventStore) {
+        this.eventSourcingSystem = new EventSourcingSystem(
+            ImmutableSet.of(
+                new ClearCommandHandler(eventStore),
+                new StoreCommandHandler(eventStore)),
+            NO_SUBSCRIBER,
+            eventStore);
+        this.eventStore = eventStore;
+    }
+
+    @Override
+    public Map<DLPRuleId, DLPRule> retrieveRules(Domain domain) {
+        return getLastEvent(domain)
```

Comment:
It's not the way it should be done.
Either you load the aggregate and ask it the list or you build a projection with a Listener and you get the list from this projection.

Reply:
Let's keep it simple and ask the aggregate the list (the aggregate will be reading the last event)